### ISSUE TITLE
SQL Server membership failure

### ISSUE THREAD

**Issue by dsarfati** [dotnet/orleans#1725]
I'm trying to get our SQL Server membership provider integrated into our project using Orleans 1.2.0-beta. I have 2 silos running in different VMs connected to a 3rd VM that is running SQL Server and MongoDB. I am able to connect only 1 silo at a time. When i try to connect the 2nd silo it runs until it gets to the membership oracle and then crashes. Here are logs when i try to run the 2nd silo:

[OrleansLog.txt](https://github.com/dotnet/orleans/files/245489/DRACOCLUSTER-1-2016-04-29-22.38.02.860Z.txt)

[StartupError.txt](https://github.com/dotnet/orleans/files/245487/DRACOCLUSTER-1-StartupError-2016-04-29-22.38.14.184Z.txt)

**Comment by gabikliot** [comment]
I think you didn't give a unique not empty name to each silo. Instance name in the log is empty. If you did, try deleting the whole sql table and try again. Maybe some empty name entry got stuck somehow in the table.

**Comment by dsarfati** [comment]
@gabikliot What is the config tag to set the instance name? I haven't seen an instance name in any of the sample configurations. Here is my config file:
[OrleansConfiguration.txt](https://github.com/dotnet/orleans/files/245653/OrleansConfiguration.txt)

**Comment by veikkoeeva** [comment]
@dsarfati Can you update tracing level to `Verbose3` and post logs. What do
   
       SELECT * FROM OrleansMembershipTable;
       SELECT * FROM OrleansMembershipVersionTable;
return?

Just to make sure, what does `SELECT QueryKey FROM OrleansQuery;` return?

**Comment by gabikliot** [comment]
@veikkoeeva , I am almost 100% sure its the empty name, and is unrelated to SQL provider.
The name is passed to Silo constructor, or any silo host.

**Comment by gabikliot** [comment]
Did you try cleaning up the table first?

**Comment by dsarfati** [comment]
Im running with OrleansHost.exe, is that how i should be doing it? The only way i can get it to run is to clean up the table and restart my silos. If it is clean, everything works ok the 1st time. Then I simulate a failure by killing the process. When i try to start the silo again i get the same error I linked above. Here is my SQL table:

![image](https://cloud.githubusercontent.com/assets/7310169/14963848/78b7cc7e-105a-11e6-8a06-3304fa4c80b1.png)

**Comment by gabikliot** [comment]
Then try what @veikkoeeva wrote.

**Comment by dsarfati** [comment]
Query:
![image](https://cloud.githubusercontent.com/assets/7310169/14964234/141410d2-105c-11e6-8136-e28361258bcb.png)

**Comment by dsarfati** [comment]
@gabikliot @veikkoeeva I'm going to try a clean run with logs Verbose3

**Comment by veikkoeeva** [comment]
@dsarfati OK. I need to be off from the computer for about half an hour now.

From the table information we can see that you have

1. One deployment: `DracoSilo`
2. Host `DRACOCLUSTER-1` is from previous month and it hasn't connected since.
3. Host `DRACODEV-1` has connected multiple times, but only one silo is alive (Status = 3). The rest are dead (Status = 6).
4. The dead ones come from the same host using the same port.

What's the IP of the other machine? Can you check, using the same connection string as in the configuration, if you can connect to the DB or is the connection blocked? It doesn't look like there is ever even an attempt to touch the membership table from any other machine than `DRACODEV-1`. Or could it be that you've cloned the VM and its network configuration didn't get updated? It shouldn't happen as easily these days, but it's a possibility too.

If you kill the process on one machine violently and restart it on the same machine with same port configuration, it might not be able to use the same port for a while as the port is in a "lingering state".

**Comment by dsarfati** [comment]
My env config is as follows:
DracoDB [192.168.143.130] SQL Server & MongoDB
DracoDev [192.168.143.134] 1 Silo, still running. Has not "crashed" yet
DracoCluster [192.168.143.129] 1 Silo. Started running fine. Killed it. Fails to start again

 Here are the logs from DracoCluster
[DRACOCLUSTER-1-StartupError-2016-05-02-20.55.30.653Z.txt](https://github.com/dotnet/orleans/files/245918/DRACOCLUSTER-1-StartupError-2016-05-02-20.55.30.653Z.txt)

![image](https://cloud.githubusercontent.com/assets/7310169/14967993/3374019a-106f-11e6-9031-9f6fa4ace7d4.png)

**Comment by veikkoeeva** [comment]
@dsarfati Not to leave you stranded here for a too long a time, it looks like other people have come across this problem too. See at https://gitter.im/dotnet/orleans?at=5728f75072798bd77be9abd8. I wonder if @shayhatsor has ideas.

**Comment by gabikliot** [comment]
I think I found the problem.
The InstanceName column in the table is missing.
I tracked it down to this PR which lost it: https://github.com/dotnet/orleans/commit/2693e5e88afb98b6fb4e84a96c7afe9ffbb06f51

We really, really, do need automated tests to run as part of CI for SQL.
Plus this particular bug may not be found by any of our SQL tests, since it only covers some but not all scenarios and removing dead silos in the table may not be covered. We need more tests to directly check data integrity.

**Comment by veikkoeeva** [comment]
@gabikliot That shouldn't be a problem performance wise, the SQL tests run very quickly (some tens of milliseconds), provided LocalDb and/or MySql are installed. Hmm, `InstanceName` missing indicates it was by mistake. I don't remember there was a reason for removing that. In which case it eluded my review. I'd like to wait for @shayhatsor to give a definite opinion.

In the meanwhile, one might succeed unblocking oneself by `ALTER TABLE OrleansMembershipTable ADD InstanceName NVARCHAR(150) NULL;` in the database. Provided there isn't something else.

**Comment by shayhatsor** [comment]
@gabikliot I totally agree, I actually added and improved many tests in the last few commits, but I also mentioned many times what you're saying. If we support something, we need to support it through and through. While Azure gets the complete tests, all others aren't. I'll be happy to help in mitigating this, we should open an issue describing the effort.
Regarding this bug, I've removed `InstanceName` because of [this](https://github.com/dotnet/orleans/blob/master/src/Orleans/SystemTargetInterfaces/IMembershipTable.cs#L243) line. The tests obviously don't cover it.

**Comment by veikkoeeva** [comment]
@shayhatsor OK. I'll leave the stage for you. Thanks for coming quickly. :)

**Comment by gabikliot** [comment]
The comment: "Azure only" is wrong at this stage. It was correct maybe at some time, but not any more. We should put InstanceName back, and it adding NULL, as @veikkoeeva  suggsted, will not fix it. It needs to be the real  silo name.

**Comment by sergeybykov** [comment]
`Silo` already has `Name`. Shouldn't it be used instead of `InstanceName`? I suspect part big part of the confusion here is that those 4 properties indeed came as Azure specific, and hence optional. But later `InstanceName` got also appropriated for a different purpose?

**Comment by shayhatsor** [comment]
@gabikliot, I've exaggerated a bit to make a point. To be concrete, we currently have a very good test coverage in terms of the 'IMembershipTable' interface. But we have a very poor test coverage in terms of tests that brings up a real silo with an arbitrary 'IMembershipTable', This is where we this bug lurked in.

**Comment by amccool** [comment]
@dsarfati by any chance are you running the OrleansDashboard?

**Comment by shayhatsor** [comment]
@sergeybykov, the `SupressDuplicateDeads` method is used in one place and only for logging purposes. It's used to get a new copy of the membership table that includes the newest dead entry for each silo and all other non-dead entries, probably to make the log message short and relevant.
You've mentioned that a silo has a 'Name', but I think it won't help here since we have it only for the local running silo. I see two possible solutions:
1) `SupressDuplicateDeads` use `InstanceName` when running on Azure and on other membership providers use `IP:Port`.
2) Always use `IP:Port`, which means that on Azure, the log might include more than one dead entry with the same `InstanceName`. 
What do you think?

**Comment by amccool** [comment]
hopefully I am not hijacking, but I think I have the same issue
use case: 1.2.0, sql membership
2 silos
if I start a single silo, start is good (3), stop is good (6)
however if a second silo is started after the first
silo1 is at **3**   the second silo errors but is at **3**
https://gist.github.com/amccool/69b08aa9f818069760583e9a89a75ef1 (at warning verbosity)
but silo2 stays running, and when ctrl+c for a terminate it still is at **3**

now silo1 is throwing loads of warning looking for silo2
(will edit and add verbose3 log for silo2)

**Comment by sergeybykov** [comment]
@shayhatsor I'm looking at it the following way. `IP:Port` part of `SiloAddress` is the authoritative identifier for a silo. `InstanceName` is just an optional friendly name for it, in case of Azure for sure, but maybe also other systems.

This makes me think that `InstanceName` should only be used for logging, and only if it's set, otherwise `IP:Port`.

Wouldn't this solve the problem?

**Comment by amccool** [comment]
@veikkoeeva 
ALTER TABLE OrleansMembershipTable add InstanceName NVARCHAR(150) NULL;
is keeping my silos ups

**Comment by amccool** [comment]
(running with InstanceName column added) upon normal shutdown of one of my three silos I got this:

```
23:45:09Z [Runtime.Scheduler.WorkerPoolThread/System.5] Error R5Ent.OrleansSiloHostService.exe 0: [2016-05-03 23:45:09.353 GMT    12    ERROR   100655  MembershipOracle        172.18.124.92:11111]
!!!!!!!!!! Error doing SiloStatusChangeNotification
Exc level 0: System.ArgumentNullException: Value cannot be null.
Parameter name: key
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Orleans.MembershipTableData.SupressDuplicateDeads()
   at Orleans.Runtime.MembershipService.MembershipOracle.<ProcessTableUpdate>d__48.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.Runtime.MembershipService.MembershipOracle.<SiloStatusChangeNotification>d__41.MoveNext()
R5Ent.OrleansSiloHostService.exe Error: 0 : [2016-05-03 23:45:09.353 GMT    12  ERROR   100655  MembershipOracle        172.18.124.92:11111]    !!!!!!!!!! Error doing SiloStatusChangeNotification

Exc level 0: System.ArgumentNullException: Value cannot be null.
Parameter name: key
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Orleans.MembershipTableData.SupressDuplicateDeads()
   at Orleans.Runtime.MembershipService.MembershipOracle.<ProcessTableUpdate>d__48.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.Runtime.MembershipService.MembershipOracle.<SiloStatusChangeNotification>d__41.MoveNext()

```

**Comment by sergeybykov** [comment]
That's the consequence of `SupressDuplicateDeads` assuming non-null `InstanceName`s.

**Comment by shayhatsor** [comment]
@sergeybykov, about your [comment](https://github.com/dotnet/orleans/issues/1725#issuecomment-216690834):
I think that checking the value of `InstanceName` in non Azure membership is not desirable. My reasoning is that it's not a part of the membership protocol, it's an Azure specific metadata. I wouldn't want anyone who's implementing a membership provider to give it a second thought. The fact the currently all other membership providers populate it with some value - isn't a good thing.

**Comment by gabikliot** [comment]
I disagre Shay. It's not azure specific. We should simply rename it to silo name. Every silo is given a name when started. In azure hosting (unrelated from azure syst store) the name comes from azure, but that is unrelated. In any hosting we should have silo name. And we do: silo host accepts a name.
I think having silo name in the MBR table is useful abd we should keep it, in all system stores.

A separate question is if we can or not use this name in MBR protocol. There is no currently reason to use it, since we can use IP:port. I will change supress dead silos to use IP: port
 But we should still keeep the silo name in the MBR table.

**Comment by shayhatsor** [comment]
@gabikliot, I see your point. InstanceName can be useful in other implementations as a meta-data. Ping me after you made the change and I'd add the silo name column to the relational storage implementation.

**Comment by gabikliot** [comment]
Exactly.

I don't have access to my windows machine to make this change right now and it will take me about a week until I will be able to fix it. It's a tiny fix, just change the suppress dead silo function to key by IP:port.

**Comment by shayhatsor** [comment]
@gabikliot, no problem, will do.

**Comment by sergeybykov** [comment]
Great. Looks like we are in agreement. 👍

**Comment by veikkoeeva** [comment]
@gabikliot. @sergeybykov Is IP:port something that just uniquely and stabily denotes a silo identifier in a given service deployment (service ID) that doubles ? If so, maybe one way to see it is that it can be something else too, depending on the situation. And then separately store IP and port information, though this *stable identifer* likely many times is IP:port. This would be to just separate concerns.

How I feel I've asked this before...

**Comment by gabikliot** [comment]
Yes, we use IP:port to uniquely identify silo. You are right, it could be something else. But by using IP:port we put no restrictions/requirement on something else (hosting infra) to provide unique identifiers, which makes it easier to host Orleans on different hosting platforms.

**Comment by sergeybykov** [comment]
@veikkoeeva IP:Port does uniquely and stably identifies a silo in a cluster (deployment). Within the scope of a deployment IP:Port is the partition key for multiple (historical) records of different 'incarnations' of the node (with different generation numbers).

InstanceId or SiloName are merely friendly names for humans looking at the table or logs. For example, in Azure PaaS a role instance may migrate to a different VM. It's IP:Port changes but the friendly name stays the same.

ServiceId has nothing to do with this though. It is for continuity across different deployments/clusters over time of the same service.

**Comment by gabikliot** [comment]
Fixed via #1728.

**Comment by veikkoeeva** [comment]
@sergeybykov That is what I wondered, if `IP:port` uniquely identifies a silo within a deployment across history, but what if the silo server is logically the same, but `IP` changes dynamically, say, between restarts?

Or if one wants to use multiple NICs (without NIC teaming) and identify the server via a host name (AD joinesd, FQDN) and let the DNS server load-balance the traffic across the two NICs. In that case it would be beneficial to identity the silo by a host name. Maybe a bit contrived as one likely would want to use teaming, at least on Windows server, but a case that occurred to me.

I'm under the impression the `IP:port` pair is not used solely to identify a silo, but also as a connection information between ones in the cluster.

**Comment by sergeybykov** [comment]
@veikkoeeva 
>That is what I wondered, if  IP:port  uniquely identifies a silo within a deployment across history, but what if the silo server is logically the same, but  IP  changes dynamically, say, between restarts?

Between restarts is fine - from the membership protocol's perspective it's a new silo anyway. If the IP address changes without a restart of the silo process, then I believe we are in trouble. However, the membership protocol should take care of it in the end - vote the silo dead, which will make it restart.

>Or if one wants to use multiple NICs (without NIC teaming) and identify the server via a host name (AD joinesd, FQDN) and let the DNS server load-balance the traffic across the two NICs. In that case it would be beneficial to identity the silo by a host name.

This is not supported today. Although, frankly, I'm not sure it's important, considering that we've not seen network bound workloads with Orleans.

>I'm under the impression the  IP:port  pair is not used solely to identify a silo, but also as a connection information between ones in the cluster.

Correct. It is used like an IP address of a machine.

**Comment by gabikliot** [comment]
Closing, as this was fixed via #1728.
